### PR TITLE
kresd: runtime fixes

### PR DESCRIPTION
**Pull request by eadwu**
###### Motivation for this change
Not sure about the directory permissions.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
Thanks for ping.  I'm not sure I understand &ndash; have you ran into some issues?  There's tmpfiles spec that should ensure existence of all three.  The one for cache was added here as well to
>  \# Ensure correct ownership in case UID or GID changes.

because of ae74a0e27.  The other two didn't exist before that commit, so they weren't added in the same way.

**Comment by vcunat** [comment]
To be clear, I see no harm in adding this, so I'm mainly curious.  (and perhaps the comments are a bit confusing in combination with these "invisible" tmpfiles)

**Comment by eadwu** [comment]
For some reason the files weren't created on my system. Kept getting CHDIR errors for (`/var/lib/knot-resolver`) and then errors about `/run/knot-resolver` not existing for the control.

**Comment by vcunat** [comment]
Oh, and `lib/systemd/` does get applied?  (from `systemd.packages = [ package ];`)

**Comment by flokli** [comment]
https://github.com/NixOS/nixpkgs/pull/95565 was merged, which should fix the `knot-resolver` tmpfiles not being applied.

Even on master, the permissions in the `kresd@` template unit differ with what's in the `tmpfiles` file. `tmpfiles` will be applied once on boot, and on each system activation, whereas systemd will apply `*DirectoryMode` when starting the service, too (thus shadowing the tmpfiles).

@vcunat I think we can use only the per-unit directives here. I sent a patch to upstream at knot-resolver@labs.nic.cz - not sure how to exacty contribute. Also opened https://gitlab.nic.cz/knot/knot-resolver/-/issues/596.

File: nixos/modules/services/networking/kresd.nix [hunk 1]
```diff
@@ -129,12 +129,18 @@ in {
...
         + "-c ${package}/lib/knot-resolver/distro-preconfig.lua -c ${configFile}";
+      # Ensure /run/knot-resolver exists
+      RuntimeDirectory = "knot-resolver";
+      RuntimeDirectoryMode = "0750";
+      # Ensure /var/lib/knot-resolver exists
+      StateDirectory = "knot-resolver";
+      StateDirectoryMode = "0750";
       # Ensure correct ownership in case UID or GID changes.
       CacheDirectory = "knot-resolver";
       CacheDirectoryMode = "0750";
```
Comment (flokli) [inline nixos/modules/services/networking/kresd.nix:140]: These permissions should all be 0770 (to be consistent with what's in the tmpfiles file)

**Comment by vcunat** [comment]
Yes, I think on NixOS the kresd service can abandon using tmpfiles in favor of these, though I can't really see any practical differences.

File: nixos/modules/services/networking/kresd.nix (hunk 1 above)
Comment (eadwu) [reply nixos/modules/services/networking/kresd.nix:140]: Done.

**Pull request merged**